File: guc/src/it/us/web/action/utenti/CambioPassword.java
/*
Copyright (C) AGPL-3.0  

This program is free software: you can redistribute it and/or modify it under the terms of the GNU Affero General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for more details.

You should have received a copy of the GNU Affero General Public License along with this program. If not, see <https://www.gnu.org/licenses/>.*/
package it.us.web.action.utenti;

import it.us.web.action.Action;
import it.us.web.action.GenericAction;
import it.us.web.bean.BGuiView;
import it.us.web.bean.BUtente;
import it.us.web.dao.GuiViewDAO;
import it.us.web.dao.UtenteDAO;
import it.us.web.exceptions.AuthorizationException;
import it.us.web.util.Md5;
import it.us.web.util.properties.Message;

public class CambioPassword extends GenericAction
{

	@Override
	public void can() throws AuthorizationException
	{
		BGuiView gui = GuiViewDAO.getView( "AMMINISTRAZIONE", "UTENTI", "CAMBIO PASSWORD" );
		can( gui, "w" );
	}

	@Override
	public void execute() throws Exception
	{
		Action returnAction = null;
		

		BUtente toUpdateUser = UtenteDAO.getUtenteBId(db, interoFromRequest( "user_id" )) ; //(BUtente) persistence.find( BUtente.class, interoFromRequest( "user_id" ) );
		
		if( validate( toUpdateUser ) )
		{
			UtenteDAO.update(db, toUpdateUser) ;
			//persistence.update( toUpdateUser );
			//persistence.commit();
	 		
 			returnAction = new Detail();
 			setMessaggio( Message.getSmart( "PASSWORD_MODIFICATA" ) );
		}
		else
		{
			setErrore( "Inserire una password di almeno 8 caratteri ed assicurarsi che sia uguale alla sua conferma" );
			returnAction = new ToCambioPassword();
		}

 		req.setAttribute( "userDetail", toUpdateUser );
		goToAction( returnAction );
	}

	private boolean validate(BUtente toUpdateUser)
	{
		boolean ret = true;
		
		String password = stringaFromRequest( "password_1" );
		String confirmPassword = stringaFromRequest( "password_2" );
		if( password == null || !password.equals( confirmPassword ) || password.length() < 8 )
		{
			ret = false;
		}
		else
		{
			toUpdateUser.setPassword( Md5.encrypt( password ) );
		}
		
		return ret;
	}


}

File: guc/WebContent/js/cambioPassword.js
/*
Copyright (C) AGPL-3.0  

This program is free software: you can redistribute it and/or modify it under the terms of the GNU Affero General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for more details.

You should have received a copy of the GNU Affero General Public License along with this program. If not, see <https://www.gnu.org/licenses/>.*/
var PWD_MIN = 10;
var PWD_MAX = 15;

  function checkFormPassword(form)
  {
	  var pwd1 = null;
	  var pwd2 = null;
	  
	  if (document.getElementById("password1")!=null)
		  pwd1 = document.getElementById("password1");
	  else
		  pwd1 = document.getElementById("password");
	  
	  pwd2 = document.getElementById("password2");
	  
    re = /^\w+$/;
    if(!re.test(pwd1.value)) {
      alert("Errore: Password contenente caratteri non supportati.");
      pwd1.focus();
      return false;
    }
    
      if(pwd1.value.length < PWD_MIN) {
        alert("Errore: la password deve essere almeno di " + PWD_MIN + " caratteri");
        pwd1.focus();
        return false;
      }
      
        if(pwd1.value.length > PWD_MAX) {
            alert("Errore: la password deve essere al massimo di " + PWD_MAX + " caratteri");
            pwd1.focus();
            return false;
          }
          
    
      re = /[0-9]/;
      if(!re.test(pwd1.value)) {
        alert("Errore: La password deve contenere almeno un numero (0-9)!");
        pwd1.focus();
        return false;
      }
      re = /[a-z]/;
      if(!re.test(pwd1.value)) {
        alert("Errore: La password deve contenere almeno una lettera minuscola (a-z)!");
        pwd1.focus();
        return false;
      }
      re = /[A-Z]/;
      if(!re.test(pwd1.value)) {
        alert("Errore: La password deve contenere almeno una lettera maiuscola (A-Z)!");
        pwd1.focus();
        return false;
    } 
  	if(pwd1.value != pwd2.value ) {
      alert("Errore: La password di conferma non corrisponde alla nuova password.");
      pwd1.focus();
      return false;
    }
 
    
  	return true;
  	
  }

   

  
  function showPolicy(){
	  alert('Policy per nuove password: \nLunghezza: 10-15 caratteri\nContenente solo lettere, numeri e underscore (_).\nContenente almeno 1 lettera minuscola (a-z)\nContenente almeno 1 lettera maiuscola (A-Z)\nContenente almeno 1 numero (0-9)');
  }
  
  function reloadOpener(){
	    window.opener.location.reload();
	    window.close();
  }
  
  function mostraNascondi(radio){
	
	  var pwd1 = null;
	  var pwd2 = null;
	  
	  if (document.getElementById("password1")!=null)
		  pwd1 = document.getElementById("password1");
	  else
		  pwd1 = document.getElementById("password");
	  
	  pwd2 = document.getElementById("password2");
	
	if (radio.id == 'female'){
		pwd1.type='text';
		pwd2.type='text';
	}
	else {
		pwd1.type='password';
		pwd2.type='password';
	}
  }
  
  

  
  function show(){
		document.getElementById("policyForm").style.display="block";
	}
	function hide(){
		document.getElementById("policyForm").style.display="none";
	}